Add caffeine metric collector

checketts opened the pull request:
This mimics the Guava cache usage for collecting metrics for https://github.com/ben-manes/caffeine

FYI @ben-manes

brian-brazil commented on simpleclient_caffeine/pom.xml:
```diff
@@ -0,0 +1,70 @@
+        <dependency>
+            <groupId>com.github.ben-manes.caffeine</groupId>
+            <artifactId>caffeine</artifactId>
+            <version>2.3.5</version>
```
Use the oldest version you can so we can ensure we're as backwards compatible as possible

checketts commented:
I was going to add it to the readme after a release has occurred. Since it wouldn't actually be available for version `0.0.20`

```
<dependency>
  <groupId>io.prometheus</groupId>
  <artifactId>simpleclient_pushgateway</artifactId>
  <version>0.0.20</version>
</dependency>
```

I also want to revisit the giant XML and replace it with a single XML snippet as an example and a text description of the various modules available and usages.

brian-brazil commented:
There'll be a release fairly soon, there's been quite a few modules added.

ben-manes commented:
Thanks! I don't think you need to support 1.0 anymore. There is also an evictionWeight metric in the 2.x line. As an FYI, the AsycLoadingCache doesn't use the Cache interface due to subtle differences, but has a synchronous() view instead. You might want to mention that in your JavaDoc.

checketts commented:
@ben-manes I switched to 2.0. Thanks for the heads up on the `AsyncLoadingCache`. Instead of calling it out in the JavaDoc, I added a method that unwraps it for the user. Think that is sufficient?

I also couldn't find the `evictionWeight ` in the `CacheStats` class. Where would I find that? Also does it behave like a gauge to indicate how likely eviction is?

ben-manes commented:
That does look better.

It looks like `evictionWeight` was added in `2.3.0`. Note that since I'm strictly following semver, you should be okay using latest releases.

checketts commented:
@ben-manes The problem we will hit if we add in the `evictionWeight` metric is that people using < `2.3` will get a method missing error if they want metrics. 

I'll see if I can still include it.

ben-manes commented:
Wouldn't the transitive dependency cause their version to upgrade?

checketts commented:
Yes, if they choose to upgrade. It is common to use `DependencyManagement` to lock in a specific version and minimize changes when adding in libraries.

For example I wouldn't want adding in a new dep like the prometheus client to upgrade my version of Jetty.

checketts commented:
I've added the CacheEviction weight and it should be backwards compatible too.

@brian-brazil  I didn't like the empty catch. Is it acceptable to add `slf4j` as a dependency for logging?

checketts commented:
I disabled the AsyncLoading test because it required Java 8 classes. I know the `prometheus_simpleclient` is intended to be compatible with Java 6, but this module is Java 8 specific (since Caffeine is). Would it be acceptable to use Java 8 for the build, and compile for Java 6?

ben-manes commented:
You may want to use AnimalSniffer if using mixed compilation targets.

brian-brazil commented on README.md:
```diff
@@ -34,6 +34,18 @@ version can be found on in the maven repository for
   <artifactId>simpleclient_pushgateway</artifactId>
   <version>0.0.20</version>
 </dependency>
+<!-- Guava Cache metrics-->
```
Don't expand this list, it's already annoying to maintain and I don't want to scare off users.

brian-brazil commented on simpleclient_caffeine/src/test/java/io/prometheus/client/cache/caffeine/CacheMetricsCollectorTest.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+
+    // Test disabled since it requires Java 8 features
```
Remove rather than commenting out.

brian-brazil commented on simpleclient_caffeine/src/main/java/io/prometheus/client/cache/caffeine/CacheMetricsCollector.java:
```diff
@@ -0,0 +1,159 @@
+        mfs.add(cacheEvictionTotal);
+
+        GaugeMetricFamily cacheEvictionWeight = new GaugeMetricFamily("caffeine_cache_eviction_weight",
+                "Cache eviction weight", labelNames);
```
Some more flavour would be good

brian-brazil commented on simpleclient_caffeine/src/main/java/io/prometheus/client/cache/caffeine/CacheMetricsCollector.java:
```diff
@@ -0,0 +1,159 @@
+
+        GaugeMetricFamily cacheEvictionWeight = new GaugeMetricFamily("caffeine_cache_eviction_weight",
+                "Cache eviction weight", labelNames);
+        mfs.add(cacheEvictionTotal);
```
cacheEvictionWeight

brian-brazil commented on simpleclient_caffeine/src/main/java/io/prometheus/client/cache/caffeine/CacheMetricsCollector.java:
```diff
@@ -0,0 +1,159 @@
+            try{
+                cacheEvictionWeight.addMetric(cacheName, stats.evictionWeight());
+            } catch (Exception e) {
+                // EvictionWeight metric is unavailable, newer version of Caffeine is needed
```
fullstop at the end of comment.

checketts commented:
I'm glad I misunderstood when you said 'it should be in the readme'. Since the comment was on the POM xaml I assumed you meant adding to the maven block. I agree that XML doesn't need to grow any more.

brian-brazil commented on README.md:
````diff
@@ -274,6 +274,28 @@ To register the log4j2 collector at root level:
+```
+
+The Caffeine equivalent is nearly identical. Again, be certain to call `recordStats()`
+ when building the cache sot hat metrics are collected.
````
so that

brian-brazil commented:
Thanks!
